### ISSUE TITLE
Monkey Health Explorer: delete mark fails WCAG AA requirement for pointer targets

### ISSUE THREAD

**Issue by eatyourgreens** [zooniverse/front-end-monorepo#7512]
## Package
- lib-classifier

## Describe the bug
The delete button for drawn marks is very small on Monkey Health Explorer, making it impossible to delete drawn marks on tablets or phones. 

<img width=400 alt="A subject image from Monkey Health Explorer, with one blue SVG circle marker added. The marker has a small X button, but the X is rendered too small to be easily tapped on a touchscreen." src=https://github.com/user-attachments/assets/adec872e-2915-47aa-a1a1-797291b296cf >

## To Reproduce

https://www.zooniverse.org/projects/mbarrierz/monkey-health-explorer/classify/workflow/14984

Try drawing a mark and deleting it with a small touchscreen. 

## Expected behavior
Buttons must have a minimum size of 24 CSS pixels. 

https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html

## Additional context
I suspect that what's happening here is that the button is being sized relative to the SVG image pixels, and these are very large images. Buttons (and interactive areas in general) should be sized in screen pixels ([CSS pixels](https://developer.mozilla.org/en-US/docs/Glossary/CSS_pixel).)

Discussions on Talk:

- Monkey Health Explorer: https://www.zooniverse.org/projects/mbarrierz/monkey-health-explorer/talk/2003/4042991
- Backyard Worlds: Binaries: https://www.zooniverse.org/projects/coolneighbors/backyard-worlds-binaries/talk/7367/4042988

The size of the button is set by the radius of `DeleteIcon` here:
https://github.com/zooniverse/front-end-monorepo/blob/85714cc71af54da6a4d0be5e11625612f41fca80/packages/lib-classifier/src/plugins/drawingTools/components/DeleteButton/DeleteButton.jsx#L31

Those sizes correspond to displayed widths/heights of 11.5 and 17.5 CSS pixels respectively.

The pointer target for `DragHandle` is a transparent circle twice the size of the visible circle. That would also work to increase the size of the pointer target for the delete button.
https://github.com/zooniverse/front-end-monorepo/blob/85714cc71af54da6a4d0be5e11625612f41fca80/packages/lib-classifier/src/plugins/drawingTools/components/DragHandle/DragHandle.jsx#L50-L51

**Comment by eatyourgreens** [comment]
I think this is a regression of an older issue, but I can't find the original bug report.

**Comment by sktw** [comment]
#7405 changed the implementation of drawing mark scaling to ensure that it is unaffected by the size of the image / viewer. Looking at the screenshot both the point mark and the delete button appear to be the correct size. Some of the drawing tool controls have responsive sizing (see #7475), making them smaller on smaller screens - the delete button only has a radius of 5px below the breakpoint.

I've tried the linked workflow on an Android phone (6.1" screen), and I'm able to successfully delete marks by pressing the delete button. Perhaps there's an issue specific to iPhone that prevents small elements receiving touch events?

**Comment by eatyourgreens** [comment]
The delete button is just too small. WCAG 2.2 level AA (which is a legal requirement) requires 24px for touch targets. 

The timing of #7405 matches up with the bug report on Talk, which says that the issue appeared a couple of months ago.

**Comment by eatyourgreens** [comment]
I took a look with Firefox dev tools. The delete button is 11.5px by 11.5px at a subject image scale of 3.18 (so 3.18 subject image pixels per displayed CSS pixel?) I'm not sure how `transform: scale(3.18)` ends up at 11.5 CSS pixels but the button's layout size is ~4.2px at a scale of 1 (measured in Firefox.)

<img width=900 alt="A blue circular mark in Firefox dev tools, at the smallest screen width. Browser dev tools are open to show the CSS layout for the delete button, which is 11.5px across." src="https://github.com/user-attachments/assets/3cea8b0e-0063-4d2a-973f-5cdd35f7138d" />

Increasing the circle radius from 5 to 8 would increase the clickable area from 11.5px to 17.5px, which still fails WCAG. However, I think this could meet the WCAG touch target size by doing what we did for other drawn marks, and adding a larger, transparent 24px circle which acts as the touch target.

**Comment by eatyourgreens** [comment]
>I've tried the linked workflow on an Android phone (6.1" screen), and I'm able to successfully delete marks by pressing the delete button. Perhaps there's an issue specific to iPhone that prevents small elements receiving touch events?

I have a problem with hand tremors, which I control with medication but it does make small touch targets impossible to use. I suspect that's the difference here, rather than a technical difference between Android and iOS. When I try to tap the delete button, more often than not I end up tapping the underlying image and creating a new mark.

**Comment by sktw** [comment]
Firefox adds the stroke width (1.5px) to the size of the circle (radius 5px), giving an overall size of 11.5px. Interestingly, Chrome ignores the stroke width in the reported size, but [apparently](https://stackoverflow.com/a/76647864) the rendering is identical.

The scale factor is calculated to exactly counteract the overall scaling (zoom + viewport), so that the rendered size of the drawing elements is just their 'quoted' dimensions in css pixels.

I agree regarding the need to make the delete button bigger. The positioning code, implemented for each mark model in `deleteButtonPosition`, would likely need to be amended to avoid excessive overlap with the mark itself. Would using a transparent buffer region increase the likelihood of the volunteer accidentally pressing delete?

**Comment by eatyourgreens** [comment]
>The scale factor is calculated to exactly counteract the overall scaling (zoom + viewport), so that the rendered size of the drawing elements is just their 'quoted' dimensions in css pixels.

Ah, right. I was assuming that `r="5"` referred to 5 pixels in the frame of the parent SVG, which I think is determined by the number of pixels in the subject image (unless there's a viewbox?)

>I agree regarding the need to make the delete button bigger. The positioning code, implemented for each mark model in deleteButtonPosition, would likely need to be amended to avoid excessive overlap with the mark itself. Would using a transparent buffer region increase the likelihood of the volunteer accidentally pressing delete?

I don't think so. My reading of [success criterion 2.5.8 (touch size)](https://www.w3.org/TR/WCAG22/#target-size-minimum) is that two different click actions (delete vs. create mark) will fail if they are less than 24px apart, so any click/tap within a 24px region centred on the delete button should be considered as a delete action. 

>Undersized targets (those less than 24 by 24 CSS pixels) are positioned so that if a 24 CSS pixel diameter circle is centred on the [bounding box](https://www.w3.org/TR/WCAG22/#dfn-bounding-boxes) of each, the circles do not intersect another target or the circle for another undersized target;

It's a little bit hard to interpret because the whole subject image canvas is a touch target, so the delete button naturally overlaps it.

**Comment by eatyourgreens** [comment]
A quick fix might be to change that radius to `r=10` or `r=11`. Here, I've changed it to 10 in the DOM inspector. This matches how the delete button used to look when I first wrote the drawing tools code (back in 2020, maybe?) It's a couple of pixels off the WCAG required size, but a lot easier to use.

<img width=900 alt="Monkey Health Explorer in desktop Firefox, with the SVG delete button size changed to r=10, so that the rendered button layout is about 22px across." src="https://github.com/user-attachments/assets/356c3353-53c2-4087-a033-f3192f1bc20b" />

**Comment by sktw** [comment]
The topmost svg element has a viewbox set to the dimensions of the subject image. This allows the viewport scaling to be calculated from `rendered_width_of_topmost_svg_element / subject_width`. 

> I don't think so. My reading of success criterion 2.5.8 (touch size) is that two different click actions (delete vs. create mark) will fail if they are less than 24px apart, so any click/tap within a 24px region centred on the delete button should be considered as a delete action.

I was meaning that if the delete button has a touch area that is bigger than its 'visual' size, a volunteer who wants to draw a new mark nearby might end up hitting the delete button instead.

**Comment by eatyourgreens** [comment]
>The topmost svg element has a viewbox set to the dimensions of the subject image. This allows the viewport scaling to be calculated from `rendered_width_of_topmost_svg_element / subject_width`.

Oh, I remember now. The SVG spec says you can ignore the SVG image scaling with [`vector-effect="non-scaling-size"`](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Attribute/vector-effect#non-scaling-stroke) but none of the browsers ever implemented that, so I added the `scale` property (intrinsic subject width / displayed width) that was dotted throughout the drawing code. When I originally added that, it was to prevent drawn marks (and buttons) from growing and shrinking as you zoomed in and out of a subject. And to keep the UI reasonably consistent across projects and workflows with a variety of image sizes.

**Comment by eatyourgreens** [comment]
>I was meaning that if the delete button has a touch area that is bigger than its 'visual' size, a volunteer who wants to draw a new mark nearby might end up hitting the delete button instead.

True. I think the line tool uses a transparent line that's twice the width of the visible line, but probably best to keep the visible control and the touch area as close as possible.

**Comment by eatyourgreens** [comment]
I originally set that button to 11px radius, presumably to make it accessible (and legal!) on small screens.
https://github.com/zooniverse/front-end-monorepo/blob/b06fe04657dac285c999a221e9359367b95db5fa/packages/lib-classifier/src/plugins/drawingTools/components/DeleteButton/DeleteButton.js#L20

I wonder why I changed it? 🤔 

I changed it to 5 here:
- #6727.

**Comment by eatyourgreens** [comment]
In the `DragHandle` drawing component, the pointer target is a transparent circle twice as big as the visible circle.
https://github.com/zooniverse/front-end-monorepo/blob/85714cc71af54da6a4d0be5e11625612f41fca80/packages/lib-classifier/src/plugins/drawingTools/components/DragHandle/DragHandle.jsx#L50-L51

That would also work to make the delete button's pointer target bigger.

**Comment by eatyourgreens** [comment]
It turns out there's also a discussion about marker size bugs in Backyard Worlds:
https://www.zooniverse.org/projects/coolneighbors/backyard-worlds-binaries/talk/7367/4042988

(added to the original issue description too.)

**Comment by eatyourgreens** [comment]
An update from Talk about small marker sizes on other projects:

> When using the marquee tool to draw boxes in the Indigo Snake Project
> 
> https://www.zooniverse.org/projects/fwc/indigo-snake-watch/talk/6126/4071744?comment=6648839
> 
> and Wild Watch Burrowing Owl Project, the cross mark to delete now appears tiny (even when the Subject image is at 100% magnification) making the boxes impossible to delete, a similar problem to the tiny target marker size in Monkey Health Explorer and other projects.
> 
> This did not used to be the case 2 months ago. Both projects used to have fully functioning marquee box drawing tools all of my devices. Both of these projects used to have deletion cross marks in the corners of the boxes with consistent sizes.
> 
> Also when the image is zoomed in, the cross marks in the right corner of the boxes shrink making them impossible to click.

https://www.zooniverse.org/projects/mbarrierz/monkey-health-explorer/talk/2003/4042991?comment=6648728

**Comment by sktw** [comment]
> Also when the image is zoomed in, the cross marks in the right corner of the boxes shrink making them impossible to click.

I've been unable to reproduce this in either Firefox (desktop) or Chrome (desktop and mobile). In all cases the delete button and drag handles remain a constant size when changing the viewer zoom level. There is a kind of optical-illusion effect where the delete button and drag handles *appear* smaller (bigger) relative to the rectangle when the image is zoomed in (out), but the pixel sizes remain constant. Perhaps the volunteer is using a different browser / platform?

Here are a couple of screenshots from the Indigo Snake Watch project on Firefox (Ubuntu desktop) showing the same rectangle zoomed in and out:

<img width="1790" height="766" alt="Image" src="https://github.com/user-attachments/assets/0514898a-768c-4946-bc51-35bd3f9e30d0" />

<img width="1806" height="758" alt="Image" src="https://github.com/user-attachments/assets/46c5bca0-6651-488d-ab9c-4346de9339bb" />

**Comment by eatyourgreens** [comment]
I think that Talk post might be referring to an Android tablet, but I'm not sure. The delete button does fall below the WCAG AA size threshold on both desktop and tablet screens.

I can't see any scaling problems, either zoomed in or out, with Firefox on MacOS.

**Comment by sktw** [comment]
The volunteer also mentioned an issue with the Iguanas from Above project. That [workflow](https://www.zooniverse.org/projects/andreavarela89/iguanas-from-above/classify/workflow/31838) uses the [point](https://github.com/zooniverse/front-end-monorepo/blob/e6afe1b22261d65d725c5fcf5d30ac66807d8451/packages/lib-classifier/src/plugins/drawingTools/components/Point/Point.jsx) tool, and specifies the `small` size in the tool config for the second task. For unselected small points the circle part is tiny, just 2px in radius. It would presumably also fail the WCAG AA size requirements.

**Comment by eatyourgreens** [comment]
The pointer target is 20px for those small marks, but that does still fail the WCAG AA criterion (24px.)

<img width="1412" height="584" alt="A screenshot of Iguanas From Above with the dev tools DOM inspector open to measure the size of a small drawn mark. It shows as 20px across." src="https://github.com/user-attachments/assets/81274ecd-8a55-4637-b240-9471c1b0023a" />

**Comment by eatyourgreens** [comment]
Iguanas From Above also has the bug where the subject image changes size when you navigate from the initial question to the drawing task, but only if 'limit subject height' is enabled.
- https://github.com/zooniverse/front-end-monorepo/issues/6275

**Comment by sktw** [comment]
> The pointer target is 20px for those small marks, but that does still fail the WCAG AA criterion (24px.)

The size of the bounding box of the parent `g` element is 20 x 20px, but the area for pointer events is defined by the sizes of the circle and crosshair lines, and is much smaller. The lines are only 1px wide, so are practically unclickable with a mouse. That just leaves the centre circle, which has a size of 6 x 6px.

**Comment by eatyourgreens** [comment]
>The size of the bounding box of the parent g element is 20 x 20px, but the area for pointer events is defined by the sizes of the circle and crosshair lines, and is much smaller. The lines are only 1px wide, so are practically unclickable with a mouse. That just leaves the centre circle, which has a size of 6 x 6px.

That's definitely a WCAG failure, then. Depending on the project owner/funding source, WCAG AA conformance can be a legal requirement for some organisations.
